src/app/components/elements/EventBookingForm.tsx

```diff
@@ -14,6 +14,112 @@ interface EventBookingFormProps {
+        School year group
+      </RS.Label>
+      <RS.Input
+        type="select"
+        id="year-group"
+        name="year-group"
+        value={additionalInformation.yearGroup || ""}
+        onChange={(event) => updateAdditionalInformation({ yearGroup: event.target.value })}
+      >
+        <option value="" />
+        {event.isAStudentEvent && (
+          <>
+            <option value="9">Year 9</option>
+            <option value="10">Year 10</option>
+            <option value="11">Year 11</option>
+            <option value="12">Year 12</option>
+            <option value="13">Year 13</option>
+          </>
+        )}
+        {!event.isStudentOnly && (
+          <>
+            <option value="TEACHER">N/A - Teacher</option>
+            <option value="OTHER">N/A - Other</option>
+          </>
+        )}
+      </RS.Input>
+      {event.isStudentOnly && (
+        <div className="text-muted" data-testid="student-only">
+          {studentOnlyEventMessage(event.id)}
+        </div>
+      )}
+      {(event.isATeacherEvent || additionalInformation.yearGroup == "TEACHER") && (
+        <div className="mt-2 text-right">
+          <a href="/pages/teacher_accounts" target="_blank">
+            Click to upgrade to a teacher account
+          </a>{" "}
+          for free!
+        </div>
+      )}
+    </>
+  );
+};
+
+export const AccessibilityAndMedicalRequirements = ({
+  additionalInformation,
+  updateAdditionalInformation,
+}: {
+  additionalInformation: AdditionalInformation;
+  updateAdditionalInformation: (update: AdditionalInformation) => void;
+}) => {
+  const data = [
+    {
+      label: "Accessibility requirements",
+      id: "access",
+      popover:
+        "For example, please let us know if you need wheelchair access, hearing loop or if we can help with any special adjustments.",
+      type: "accessibilityRequirements",
+    },
+    {
+      label: "Dietary requirements or relevant medical conditions",
+      id: "medical",
+      popover:
+        "For example, please let us know if you need wheelchair access, hearing loop or if we can help with any special adjustments.",
+      type: "medicalRequirements",
+    },
+  ];
+
+  return (
+    <div>
+      {data.map((each) => {
+        const { label, id, popover, type } = each;
+        return (
+          <div key={id}>
+            <RS.Label htmlFor={`${id}-reqs`}>
+              {label}
+              <span id={`${id}-reqs-help`} aria-haspopup="true" className="icon-help has-tip" />
+              <RS.UncontrolledPopover trigger="click" placement="bottom" target={`${id}-reqs-help`}>
+                <RS.PopoverBody>{popover}</RS.PopoverBody>
+              </RS.UncontrolledPopover>
+            </RS.Label>
+            <RS.Input
+              id={`${id}-reqs`}
+              name={`${id}-reqs`}
+              type="text"
+              value={additionalInformation[type as keyof AdditionalInformation] || ""}
```

Comment:
For my own benefit: what is `as keyof` doing here? I've never come across this, and I would have thought that `additionalInformation[type]` should have been enough.

Reply:
Without that it gives this:

```
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'AdditionalInformation'.
  No index signature with a parameter of type 'string' was found on type 'AdditionalInformation'.ts(7053)
```

Essentially because additionalInformation is typed, it needs to know that you're giving it an actual property of it, not a random string? I guess an alternative way of handling it would be to supply the `type` inside the data object as a keyof Additional Information rather than a string?

Reply:
Yeah, that's normally the way to go. Since `data` is a variable that you have created and therefore can fully control, you can define the exact types of its attributes in the declaration (implicitly or explicitly). That way you don't have to go doing castings like this later on. In this example it's not a big deal, but just bear in mind that castings shouldn't be a substitute of good type choices, it's better to define things properly on the first instance, if that makes sense. So in this case it looks like you have implicitly declared `type` as `string` (incorrectly), but it should have been `keyof AdditionalInformation` (if it's even possible to declare a type like that, which may not be, in which case the string casting may be the right thing to do after all).